futaba/cogs/optional/gist/core.py

```diff
@@ -0,0 +1,249 @@
+
+    def change_setting(self, guild, setting, value):
+        settings = self.get_settings(guild)
+
+        if len(settings) == 0:
+            settings = self.default_settings
+
+        settings[setting] = value
+
+        self.set_settings(guild, settings)
+
+    @staticmethod
+    def format_message_contents(message):
+        contents = message.content.split("\n")
+        contents = "  \n".join(">" + content for content in contents)
+
+        return f"{user_discrim(message.author)}:  \n{contents}  \n  \n"
+
+    @commands.command(name="gist", aliases=["msgupload"])
+    @commands.guild_only()
+    async def upload_message(self, ctx, *messages: MessageConv):
+        """
+        Concatenates the range of messages and upload to a gist.
+        A link to the gist is posted after a successful upload.
+        """
+        if len(messages) == 0:
+            raise CommandFailed(
+                content="Please specify the messages that should be uploaded."
+            )
+
+        settings = self.get_settings(ctx.guild)
+        oauth_token = settings.get("token")
+
+        if not oauth_token:
+            raise CommandFailed(content="The gist oauth token is not configured.")
+
+        messages_content = "\n".join(
+            self.format_message_contents(message) for message in messages
+        )
+        messages_ids = ", ".join(str(message.id) for message in messages)
+
+        gist_url = await create_single_gist(
+            token=oauth_token,
+            content=messages_content,
+            filename=settings.get("filename"),
+            description=settings.get("description"),
+            public=settings.get("public"),
+        )
+
+        logger.info(
+            "Successfully uploaded %d messages[%s] into a gist. Requested by user '%s' (id=%d, guild=%d)",
+            len(messages),
+            messages_ids,
+            ctx.author.name,
+            ctx.author.id,
+            ctx.guild.id,
+        )
+
+        embed = discord.Embed(description="Done! Messages successfully uploaded!")
+        embed.add_field(name="Permalink", value=gist_url)
+        embed.colour = discord.Colour.dark_teal()
+
+        await ctx.send(embed=embed)
+
+    @commands.command(name="mvgist", aliases=["msgcollapse"])
+    @commands.guild_only()
+    async def collapse_message(self, ctx, *messages: MessageConv):
+        """
+        Concatenates the range of messages and uploads to a gist.
+        The original messages are deleted and a link to the gist is posted.
+
+        Note: The messages specified should be by the same user
+        """
+
+        if not permissions.has_perm(ctx, "manage_messages") and any(
+            message.author.id != ctx.author.id for message in messages
+        ):
+            # check if the messages were created by the same user
+            raise ManualCheckFailure(content="I can only collapse your messages")
+
+        await self.upload_message(ctx, *messages)
```

Comment:
Would prefer if this was a common helper method instead of directly invoking another command. If arguments or things involving pre/post command hooks change then this will cause issues.